The close() method does not reliably close the SSE connection

Consider the following example:

```ts
let sse = new EventSource('https://httpstat.us/404', {
    debug: true,
    pollingInterval: 10000,
});

sse.addEventListener('error', (e) => {
    console.log('There was an error:', e);
    console.log('Closing SSE');
    sse.close();
});
```

The `httpstat.us/404` endpoint always returns a 404 error status. Since I call `sse.close()` in the error handler, I would expect the SSE connection be closed pretty much immediately after it is created. However, this is not the case. If I run the above code, the SSE keeps trying to reconnect every 10 seconds . Apparently, calling `sse.close()` does not properly stop the pollingInterval. Here are the logs I get when running the above code (it keeps repeating every 10 seconds):

```
 DEBUG  [EventSource][onreadystatechange] ReadyState: 2, status: 404
 DEBUG  [EventSource][onreadystatechange] ReadyState: 3, status: 404
 LOG  There was an error: {"message": "404 Not Found", "type": "error", "xhrState": 3, "xhrStatus": 404}
 LOG  Closing SSE
 DEBUG  [EventSource][onreadystatechange] ReadyState: 4, status: 0
 DEBUG  [EventSource][onreadystatechange][ERROR] Response status error. Reconnecting...
 DEBUG  [EventSource][onreadystatechange] ReadyState: 2, status: 404
 DEBUG  [EventSource][onreadystatechange] ReadyState: 3, status: 404
 LOG  There was an error: {"message": "404 Not Found", "type": "error", "xhrState": 3, "xhrStatus": 404}
 LOG  Closing SSE
 DEBUG  [EventSource][onreadystatechange] ReadyState: 4, status: 0
 DEBUG  [EventSource][onreadystatechange][ERROR] Response status error. Reconnecting...
 DEBUG  [EventSource][onreadystatechange] ReadyState: 2, status: 404
 DEBUG  [EventSource][onreadystatechange] ReadyState: 3, status: 404
 LOG  There was an error: {"message": "404 Not Found", "type": "error", "xhrState": 3, "xhrStatus": 404}
 LOG  Closing SSE
 DEBUG  [EventSource][onreadystatechange] ReadyState: 4, status: 0
 DEBUG  [EventSource][onreadystatechange][ERROR] Response status error. Reconnecting...
```

If I change the error handler to only close the SSE after waiting one second (using `setTimeout`), then it works:

```ts
let sse = new EventSource('https://httpstat.us/404', {
    debug: true,
    pollingInterval: 10000,
});

sse.addEventListener('error', (e) => {
    console.log('There was an error:', e);
    setTimeout(() => {
        console.log('Closing SSE');
        sse.close();
    }, 1000);
});
```

The 404 error is yielded twice for some reason, but this time closing the connection works:

```
 DEBUG  [EventSource][onreadystatechange] ReadyState: 2, status: 404
 DEBUG  [EventSource][onreadystatechange] ReadyState: 3, status: 404
 LOG  There was an error: {"message": "404 Not Found", "type": "error", "xhrState": 3, "xhrStatus": 404}
 DEBUG  [EventSource][onreadystatechange] ReadyState: 4, status: 404
 LOG  There was an error: {"message": "404 Not Found", "type": "error", "xhrState": 4, "xhrStatus": 404}
 DEBUG  [EventSource][onreadystatechange][ERROR] Response status error. Reconnecting...
 LOG  Closing SSE
 LOG  Closing SSE
```

In my opinion, this is a pretty serious bug. The SSE should always stop all its activity when the `close()` method is called. It certainly shouldn't try to auto-reconnect after this method was called.

--- a/README.md
+++ b/README.md
@@ -188,6 +188,7 @@ const options: EventSourceOptions = {
   method: 'GET'; // Request method. Default: GET
   timeout: 0; // Time (ms) after which the connection will expire without any activity. Default: 0 (no timeout)
   timeoutBeforeConnection: 500; // Time (ms) to wait before initial connection is made. Default: 500
+  withCredentials: false; // Include credentials in cross-site Access-Control requests. Default: false
   headers: {}; // Your request headers. Default: {}
   body: undefined; // Your request body sent on connection. Default: undefined
   debug: false; // Show console.debug messages for debugging purpose. Default: false

--- a/README.md
+++ b/README.md
@@ -186,10 +186,10 @@ new EventSource(url: string | URL, options?: EventSourceOptions);
 ```typescript
 const options: EventSourceOptions = {
   method: 'GET'; // Request method. Default: GET
-  timeout: 0; // Time after which the connection will expire without any activity: Default: 0 (no timeout)
-  timeoutBeforeConnection: 500; // Time to wait before initial connection is made: Default: 500ms
+  timeout: 0; // Time (ms) after which the connection will expire without any activity. Default: 0 (no timeout)
+  timeoutBeforeConnection: 500; // Time (ms) to wait before initial connection is made. Default: 500
   headers: {}; // Your request headers. Default: {}
-  body: undefined; // Your request body sent on connection: Default: undefined
+  body: undefined; // Your request body sent on connection. Default: undefined
   debug: false; // Show console.debug messages for debugging purpose. Default: false
   pollingInterval: 5000; // Time (ms) between reconnections. Default: 5000
 }

--- a/README.md
+++ b/README.md
@@ -191,7 +191,7 @@ const options: EventSourceOptions = {
   headers: {}; // Your request headers. Default: {}
   body: undefined; // Your request body sent on connection. Default: undefined
   debug: false; // Show console.debug messages for debugging purpose. Default: false
-  pollingInterval: 5000; // Time (ms) between reconnections. Default: 5000
+  pollingInterval: 5000; // Time (ms) between reconnections. If set to 0, reconnections will be disabled. Default: 5000
 }
 ```
 

--- a/README.md
+++ b/README.md
@@ -185,14 +185,14 @@ new EventSource(url: string | URL, options?: EventSourceOptions);
 
 ```typescript
 const options: EventSourceOptions = {
-  method: 'GET'; // Request method. Default: GET
-  timeout: 0; // Time (ms) after which the connection will expire without any activity. Default: 0 (no timeout)
-  timeoutBeforeConnection: 500; // Time (ms) to wait before initial connection is made. Default: 500
-  withCredentials: false; // Include credentials in cross-site Access-Control requests. Default: false
-  headers: {}; // Your request headers. Default: {}
-  body: undefined; // Your request body sent on connection. Default: undefined
-  debug: false; // Show console.debug messages for debugging purpose. Default: false
-  pollingInterval: 5000; // Time (ms) between reconnections. If set to 0, reconnections will be disabled. Default: 5000
+  method: 'GET', // Request method. Default: GET
+  timeout: 0, // Time (ms) after which the connection will expire without any activity. Default: 0 (no timeout)
+  timeoutBeforeConnection: 500, // Time (ms) to wait before initial connection is made. Default: 500
+  withCredentials: false, // Include credentials in cross-site Access-Control requests. Default: false
+  headers: {}, // Your request headers. Default: {}
+  body: undefined, // Your request body sent on connection. Default: undefined
+  debug: false, // Show console.debug messages for debugging purpose. Default: false
+  pollingInterval: 5000, // Time (ms) between reconnections. If set to 0, reconnections will be disabled. Default: 5000
 }
 ```
 

--- a/index.d.ts
+++ b/index.d.ts
@@ -47,6 +47,7 @@ export interface EventSourceOptions {
   method?: string;
   timeout?: number;
   timeoutBeforeConnection?: number;
+  withCredentials?: boolean;
   headers?: Record<string, any>;
   body?: any;
   debug?: boolean;

--- a/index.d.ts
+++ b/index.d.ts
@@ -46,11 +46,11 @@ export interface ExceptionEvent {
 export interface EventSourceOptions {
   method?: string;
   timeout?: number;
+  timeoutBeforeConnection?: number;
   headers?: Record<string, any>;
   body?: any;
   debug?: boolean;
   pollingInterval?: number;
-  timeoutBeforeConnection?: number;
 }
 
 export type EventSourceEvent = MessageEvent | OpenEvent | CloseEvent | TimeoutEvent | ErrorEvent | ExceptionEvent;

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -132,10 +132,7 @@ class EventSource {
       if (this.timeout > 0) {
         setTimeout(() => {
           if (this._xhr.readyState === XMLHttpRequest.LOADING) {
-            this.dispatch('error', {
-              type: 'timeout',
-            });
-
+            this.dispatch('error', { type: 'timeout' });
             this.close();
           }
         }, this.timeout);

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -75,6 +75,10 @@ class EventSource {
       this._xhr.timeout = this.timeout;
 
       this._xhr.onreadystatechange = () => {
+        if (this.status === this.CLOSED) {
+          return;
+        }
+
         const xhr = this._xhr;
 
         this._logDebug(`[EventSource][onreadystatechange] ReadyState: ${xhr.readyState}, status: ${xhr.status}`);
@@ -95,16 +99,14 @@ class EventSource {
             this._logDebug('[EventSource][onreadystatechange][DONE] Operation done.');
             this._pollAgain(this.interval, false);
           }
-        } else if (this.status !== this.CLOSED) {
-          if (this._xhr.status !== 0) {
-            this.status = this.ERROR;
-            this.dispatch('error', {
-              type: 'error',
-              message: xhr.responseText,
-              xhrStatus: xhr.status,
-              xhrState: xhr.readyState,
-            });
-          }
+        } else if (xhr.status !== 0) {
+          this.status = this.ERROR;
+          this.dispatch('error', {
+            type: 'error',
+            message: xhr.responseText,
+            xhrStatus: xhr.status,
+            xhrState: xhr.readyState,
+          });
 
           if (xhr.readyState === XMLHttpRequest.DONE) {
             this._logDebug('[EventSource][onreadystatechange][ERROR] Response status error.');
@@ -114,6 +116,10 @@ class EventSource {
       };
 
       this._xhr.onerror = () => {
+        if (this.status === this.CLOSED) {
+          return;
+        }
+
         this.status = this.ERROR;
         this.dispatch('error', {
           type: 'error',

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -20,6 +20,7 @@ class EventSource {
     this.method = options.method || 'GET';
     this.timeout = options.timeout ?? 0;
     this.timeoutBeforeConnection = options.timeoutBeforeConnection ?? 500;
+    this.withCredentials = options.withCredentials || false;
     this.headers = options.headers || {};
     this.body = options.body || undefined;
     this.debug = options.debug || false;
@@ -57,6 +58,10 @@ class EventSource {
 
       this._xhr = new XMLHttpRequest();
       this._xhr.open(this.method, this.url, true);
+
+      if (this.withCredentials) {
+        this._xhr.withCredentials = true;
+      }
 
       if (this.headers) {
         for (const [key, value] of Object.entries(this.headers)) {

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -198,7 +198,7 @@ class EventSource {
     if (this.eventHandlers[type] === undefined) {
       this.eventHandlers[type] = [];
     }
-    
+
     this.eventHandlers[type].push(listener);
   }
 

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -97,6 +97,7 @@ class EventSource {
           }
         } else if (this.status !== this.CLOSED) {
           if (this._xhr.status !== 0) {
+            this.status = this.ERROR;
             this.dispatch('error', {
               type: 'error',
               message: xhr.responseText,
@@ -112,9 +113,8 @@ class EventSource {
         }
       };
 
-      this._xhr.onerror = (e) => {
-        this.status === this.ERROR;
-
+      this._xhr.onerror = () => {
+        this.status = this.ERROR;
         this.dispatch('error', {
           type: 'error',
           message: this._xhr.responseText,

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -18,12 +18,12 @@ class EventSource {
     };
 
     this.method = options.method || 'GET';
-    this.timeout = options.timeout || 0;
+    this.timeout = options.timeout ?? 0;
     this.timeoutBeforeConnection = options.timeoutBeforeConnection ?? 500;
     this.headers = options.headers || {};
     this.body = options.body || undefined;
     this.debug = options.debug || false;
-    this.interval = options.pollingInterval || 5000;
+    this.interval = options.pollingInterval ?? 5000;
 
     this._xhr = null;
     this._pollTimer = null;

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -38,13 +38,16 @@ class EventSource {
       this.url = url;
     }
 
-    this._pollAgain(this.timeoutBeforeConnection);
+    this._pollAgain(this.timeoutBeforeConnection, true);
   }
 
-  _pollAgain(time) {
-    this._pollTimer = setTimeout(() => {
-      this.open();
-    }, time);
+  _pollAgain(time, allowZero) {
+    if (time > 0 || allowZero) {
+      this._logDebug(`[EventSource] Will open new connection in ${time} ms.`);
+      this._pollTimer = setTimeout(() => {
+        this.open();
+      }, time);
+    }
   }
 
   open() {
@@ -89,8 +92,8 @@ class EventSource {
           this._handleEvent(xhr.responseText || '');
 
           if (xhr.readyState === XMLHttpRequest.DONE) {
-            this._logDebug('[EventSource][onreadystatechange][DONE] Operation done. Reconnecting...');
-            this._pollAgain(this.interval);
+            this._logDebug('[EventSource][onreadystatechange][DONE] Operation done.');
+            this._pollAgain(this.interval, false);
           }
         } else if (this.status !== this.CLOSED) {
           if (this._xhr.status !== 0) {
@@ -103,8 +106,8 @@ class EventSource {
           }
 
           if ([XMLHttpRequest.DONE, XMLHttpRequest.UNSENT].includes(xhr.readyState)) {
-            this._logDebug('[EventSource][onreadystatechange][ERROR] Response status error. Reconnecting...');
-            this._pollAgain(this.interval);
+            this._logDebug('[EventSource][onreadystatechange][ERROR] Response status error.');
+            this._pollAgain(this.interval, false);
           }
         }
       };

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -5,7 +5,6 @@ class EventSource {
   CLOSED = 2;
 
   constructor(url, options = {}) {
-    this.interval = options.pollingInterval || 5000;
     this.lastEventId = null;
     this.lastIndexProcessed = 0;
     this.eventType = undefined;
@@ -20,10 +19,11 @@ class EventSource {
 
     this.method = options.method || 'GET';
     this.timeout = options.timeout || 0;
+    this.timeoutBeforeConnection = options.timeoutBeforeConnection ?? 500;
     this.headers = options.headers || {};
     this.body = options.body || undefined;
     this.debug = options.debug || false;
-    this.timeoutBeforeConnection = options.timeoutBeforeConnection ?? 500;
+    this.interval = options.pollingInterval || 5000;
 
     this._xhr = null;
     this._pollTimer = null;

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -106,7 +106,7 @@ class EventSource {
             });
           }
 
-          if ([XMLHttpRequest.DONE, XMLHttpRequest.UNSENT].includes(xhr.readyState)) {
+          if (xhr.readyState === XMLHttpRequest.DONE) {
             this._logDebug('[EventSource][onreadystatechange][ERROR] Response status error.');
             this._pollAgain(this.interval, false);
           }

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -19,7 +19,7 @@ class EventSource {
     };
 
     this.method = options.method || 'GET';
-    this.timeout = options.timeOut || 0;
+    this.timeout = options.timeout || 0;
     this.headers = options.headers || {};
     this.body = options.body || undefined;
     this.debug = options.debug || false;

--- a/src/EventSource.js
+++ b/src/EventSource.js
@@ -74,11 +74,7 @@ class EventSource {
       this._xhr.onreadystatechange = () => {
         const xhr = this._xhr;
 
-        if (this.debug) {
-          console.debug(
-            `[EventSource][onreadystatechange] ReadyState: ${xhr.readyState}, status: ${xhr.status}`
-          );
-        }
+        this._logDebug(`[EventSource][onreadystatechange] ReadyState: ${xhr.readyState}, status: ${xhr.status}`);
 
         if (![XMLHttpRequest.DONE, XMLHttpRequest.LOADING].includes(xhr.readyState)) {
           return;
@@ -93,11 +89,7 @@ class EventSource {
           this._handleEvent(xhr.responseText || '');
 
           if (xhr.readyState === XMLHttpRequest.DONE) {
-            if (this.debug) {
-              console.debug(
-                '[EventSource][onreadystatechange][DONE] Operation done. Reconnecting...'
-              );
-            }
+            this._logDebug('[EventSource][onreadystatechange][DONE] Operation done. Reconnecting...');
             this._pollAgain(this.interval);
           }
         } else if (this.status !== this.CLOSED) {
@@ -111,12 +103,7 @@ class EventSource {
           }
 
           if ([XMLHttpRequest.DONE, XMLHttpRequest.UNSENT].includes(xhr.readyState)) {
-            if (this.debug) {
-              console.debug(
-                '[EventSource][onreadystatechange][ERROR] Response status error. Reconnecting...'
-              );
-            }
-
+            this._logDebug('[EventSource][onreadystatechange][ERROR] Response status error. Reconnecting...');
             this._pollAgain(this.interval);
           }
         }
@@ -157,6 +144,12 @@ class EventSource {
         message: e.message,
         error: e,
       });
+    }
+  }
+
+  _logDebug(...msg) {
+    if (this.debug) {
+      console.debug(...msg);
     }
   }
 

